### PR TITLE
Mtf finetuning

### PR DESCRIPTION
**Pull request by lintangsutawika**
This is the multitask-finetuning part @thomasw21

### CODE REVIEW

File: train_ND_MTF_gpt.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
+# coding=utf-8
```
Comment (thomasw21) [inline train_ND_MTF_gpt.py:1]: I'd rather have this in another PR. I really only want to merge `non_causal_mtf_dataset.py` in this PR + test on that dataloder, if that's alright.

Comment (lintangsutawika) [reply train_ND_MTF_gpt.py]: Right, i've removed scripts besides `non_causal_mtf_dataset.py` and reverted `arguments.py` and `tokenizer.py`

File: tests/data/t0/ag_news_classify_question_first.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
+{"text": "What label best describes this news article?\nWall St. Bears Claw Back Into the Black (Reuters) Reuters - Short-sellers, Wall Street's dwindling\\band of ultra-cynics, are seeing green again.<|endoftext|>Business", "inputs": "What label best describes this news article?\nWall St. Bears Claw Back Into the Black (Reuters) Reuters - Short-sellers, Wall Street's dwindling\\band of ultra-cynics, are seeing green again.", "targets": "Business"}
```
Comment (thomasw21) [inline tests/data/t0/ag_news_classify_question_first.json:1]: Feel free to upload this dataset on the hub, and load from there.

Comment (thomasw21) [reply tests/data/t0/ag_news_classify_question_first.json:1]: I also think you should just keep `inputs` and `targets` and tweak the preprocessing script in order to return `{tokenized_inputs} {impossible_token_id} {tokenized_targets}`.

File: megatron/data/data_samplers.py [hunk 3]
```diff
@@ -158,3 +166,59 @@ def __iter__(self):
...
+        g.manual_seed(self.epoch)
+
+        random_idx = torch.randperm(bucket_size, generator=g).tolist()
+        idx_range = [start_idx + x for x in random_idx[bucket_offset:]]
+
+        batch = []
+        # Last batch if not complete will be dropped.
+        for idx in idx_range:
+            batch.append(idx)
+            if len(batch) == self.micro_batch_size:
```
Comment (thomasw21) [inline megatron/data/data_samplers.py:221]: I'm guessing the packing changes this condition, ie you need some sort of access to the sequence length of each sample.

File: megatron/data/non_causal_mtf_dataset.py [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+            print_rank_0(' > elasped time to build and save doc-idx mapping '
+                         '(seconds): {:4f}'.format(time.time() - start_time))
+            # sample-idx.
+            start_time = time.time()
+            # Use C++ implementation for speed.
+            # First compile and then import.
+            from megatron.data import helpers
+            assert doc_idx.dtype == np.int32
+            assert sizes.dtype == np.int32
+            sample_idx = helpers.build_sample_idx(sizes, doc_idx, seq_length,
```
Comment (thomasw21) [inline megatron/data/non_causal_mtf_dataset.py:366]: I don't think we want this as documents now vary in size. So now you probably want to truncate to the right is samples are too big instead of wrapping them.

File: megatron/data/non_causal_mtf_dataset.py [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+            # shuffle-idx.
+            start_time = time.time()
+            # -1 is due to data structure used to retieve the index:
+            #    sample i --> [sample_idx[i], sample_idx[i+1])
+            if separate_last_epoch:
+                num_samples_ = num_samples_from_epochs_minus_one
+            else:
+                num_samples_ = sample_idx.shape[0] - 1
+            shuffle_idx = _build_shuffle_idx(num_samples_,
+                                             sample_idx.shape[0] - 1, np_rng)
```
Comment (thomasw21) [inline megatron/data/non_causal_mtf_dataset.py:381]: We should have a document shuffling.

File: megatron/data/non_causal_mtf_dataset.py [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+    def __getitem__(self, idx):
+        # Get the shuffled index.
+        idx = self.shuffle_idx[idx]
+        doc_idx = self.sample_idx[idx][0]
+
+        sample = self.indexed_dataset.get(
+            self.doc_idx[doc_idx]
+        )
+
+        eod_idx = np.where(sample == self.tokenizer.eod)[0]
```
Comment (thomasw21) [inline megatron/data/non_causal_mtf_dataset.py:273]: Yeah I really want to change this condition to something else ...

File: megatron/data/non_causal_mtf_dataset.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+        return self.sample_idx.shape[0] - 1
+
+    def __getitem__(self, idx):
+        # Get the shuffled index.
+        idx = self.shuffle_idx[idx]
+        doc_idx = self.sample_idx[idx][0]
+
+        sample = self.indexed_dataset.get(
+            self.doc_idx[doc_idx]
+        )
```
Comment (thomasw21) [inline megatron/data/non_causal_mtf_dataset.py:271]:
```suggestion
        sample = self.indexed_dataset.get(
            self.doc_idx[idx]
        )
```
Shouldn't it be this instead?

File: megatron/data/non_causal_mtf_dataset.py [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+
+    # Filename of the index mappings.
+    _filename = data_prefix
+    _filename += '_{}_indexmap'.format(name)
+    _filename += '_{}ns'.format(num_samples)
+    _filename += '_{}sl'.format(seq_length)
+    _filename += '_{}s'.format(seed)
+    doc_idx_filename = _filename + '_doc_idx.npy'
+    sample_idx_filename = _filename + '_sample_idx.npy'
+    shuffle_idx_filename = _filename + '_shuffle_idx.npy'
```
Comment (thomasw21) [inline megatron/data/non_causal_mtf_dataset.py:305]: I'd suggest adding a `mtf` mention to not overwrite otherdatasets.

**Review by thomasw21 — COMMENTED** [review]
Very high level review, I think it sounds like a good plan currently. Let me know if you need help implementing things.

File: megatron/data/non_causal_mtf_dataset.py (hunk 8 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mtf_dataset.py:303]: Something like this maybe?
```
doc_idx_filename = _filename + '_mtf_doc_idx.npy'
sample_idx_filename = _filename + '_mtf_sample_idx.npy'
shuffle_idx_filename = _filename + '_mtf_shuffle_idx.npy'
```

File: tests/data/t0/gpt2-tiny-merges.txt [hunk 9]
```diff
@@ -0,0 +1,4744 @@
+#version: 0.2 - Trained by `huggingface/tokenizers`
```
[the remark this answers is not in the document]
Comment (lintangsutawika) [reply tests/data/t0/gpt2-tiny-merges.txt]: Removed

File: megatron/data/data_samplers.py (hunk 3 above)
Comment (lintangsutawika) [reply megatron/data/data_samplers.py:222]: Yes, I think I will need to add the dataset directly into the data loader.

File: megatron/data/non_causal_mtf_dataset.py (hunk 4 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mtf_dataset.py:364]: Should I replace 
```
# Use C++ implementation for speed.
# First compile and then import.
from megatron.data import helpers
assert doc_idx.dtype == np.int32
assert sizes.dtype == np.int32
sample_idx = helpers.build_sample_idx(sizes, doc_idx, seq_length,
    num_epochs, tokens_per_epoch)
```
With this?
```
sample_idx = random.sample(doc_idx, len(doc_idx))
```

File: megatron/data/non_causal_mtf_dataset.py (hunk 5 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mtf_dataset.py:379]: Should I replace it with this?
```
sample_idx = random.sample(doc_idx, len(doc_idx))
```

File: megatron/data/non_causal_mtf_dataset.py (hunk 6 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mtf_dataset.py]: I changed it to 
```
        eod_idx = np.where(sample == self.impossible_token)[0]
```

It doesn't need tokenizer anymore.

File: pretrain_t0.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    #                             eval(f"args.{s}_weighted_split_splits"),
+    #                             eval(f"args.{s}_weighted_split_names"))
+    #         for paths, weights, splits, name in data_groups:
+    #             d = build_dataset_group(name, paths, weights, splits,
+    #                                     args.data_impl,
+    #                                     train_val_test_num_samples,
+    #                                     args.seq_length, args.seed,
+    #                                     (not args.mmap_warmup),
+    #                                     train_valid_test=s)
+    #             eval(f"{s}_ds").append(d)
```
Comment (Muennighoff) [inline pretrain_t0.py:196]: I think we can also support this one, no?
This is the one we used for the 176B training & probably the one we want to use for MTF

Comment (thomasw21) [reply pretrain_t0.py:196]: Dunno, we probably want a single dataset and not play with weights ... why do you think we need to have weights?

File: examples/finetune_t0.sh [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+# Runs the "345M" parameter model
+
+RANK=0
+WORLD_SIZE=1
+
+DATA_PATH="{ \
+        'input_tokens': 'tests/data/t0/ag_news_prompt_inputs_document',  \
+       'target_tokens': 'tests/data/t0/ag_news_prompt_targets_document'  \
+       }"
```
Comment (Muennighoff) [inline examples/finetune_t0.sh:11]: What about expecting the user to provide the data in the form
`input_document target_document input_document2 target_document2...`?
And then we turn it into `[[input_doc, target_doc], [input_doc2, target_doc2] ...`.
Will open a PR for this if you agree

Comment (lintangsutawika) [reply examples/finetune_t0.sh:11]: Not sure. I think it's important we future-proof it, but currently for Bloom-Zero the above should suffice. 

Also, what about this?
```
DATA_PATH="{ \
        'input_tokens': ['input_document_0', 'input_document_1', ...] ,  \
       'target_tokens': ['target_document_0', 'target_document_1', ...]  \
       }"
```

Comment (Muennighoff) [reply examples/finetune_t0.sh:11]: The problem is `--data-path` turns the above into a list based on spaces, i.e. `["{", "'input_tokens':", ..]`, so we would need a new argument that leaves it as a string & then load this in via json.

Comment (lintangsutawika) [reply examples/finetune_t0.sh:11]: Hmmm, I see, my understanding was that it would be a single string line which can be parsed like this
```
    if args.data_path:

        import json
        data_path_dict = [json.loads(args.data_path)]

        train_ds, valid_ds, test_ds = build_train_valid_test_datasets(
            data_prefix=args.data_path,
            data_impl=args.data_impl,
            splits_string=args.split,
            train_valid_test_num_samples=train_val_test_num_samples,
            seq_length=args.seq_length,
            seed=args.seed,
            skip_warmup=(not args.mmap_warmup))
```

Comment (thomasw21) [reply examples/finetune_t0.sh:11]: yeah I think the API is much cleaner. you can alway have the same prefix and have the dataset find "_inputs" and "_targets" as suffix. (in GPTDataset you'll see that there's a bunch of files)

**Review by thomasw21 — CHANGES_REQUESTED** [review]
Blocking so we don't accidentally merge this into `master`. I'm working on another branch and slowly incorporating the changes.

**Comment by lintangsutawika** [comment]
Should I change the merge direction to another branch instead of main?

**Comment by thomasw21** [comment]
No it's all good.

**Comment by thomasw21** [comment]
Closing as we've merged the other branches
